### ISSUE TITLE
[Design] Expose native GAFIME kernels as first-class Python execution surfaces (CuPy/Torch/CPU)

### ISSUE THREAD

**Issue by onlyxItachi** [onlyxItachi/GAFIME#99]
## Motivation

GAFIME already has native execution backends and memory interoperability paths, but Python frameworks should be able to connect to GAFIME at the **function level**, not only through Arrow / array-memory interchange.

The immediate use case is cloud/notebook execution (especially Colab): a user should be able to import GAFIME, hand it an existing CuPy/Torch device array, and invoke GAFIME's real native kernels without building a second implementation in Python or forcing the user through C++ / `torch.utils.cpp_extension`.

This should be an **execution/interop surface**, not a new compute backend.

## Direction

Expose stable Python-callable handles for selected native GAFIME operations, with explicit backend and execution-precision namespaces.

Illustrative shape (API naming is intentionally not frozen):

```python
gafime.backend.cupy.fp32.MI_hist(...)
gafime.backend.torch.fp32.MI_hist(...)
gafime.backend.cpu.fp32.MI_hist(...)
```

or an equivalent exported-kernel object/function interface.

For CuPy specifically, the preferred implementation should reuse GAFIME's existing CUDA kernels (e.g. PTX/CUBIN/module/function handles or an equivalent thin binding) rather than reimplementing algorithms as separate CuPy kernels. A CuPy-facing wrapper may expose RawKernel/RawModule-compatible execution where useful, but the source of truth remains the native GAFIME CUDA implementation.

## Goals

- **Zero-copy / low-copy interoperability** with existing framework-owned buffers where possible.
  - CuPy: `cupy.ndarray`, `__cuda_array_interface__`, DLPack where appropriate.
  - Torch: CUDA tensors through a user-facing Python call path with no requirement for users to write C++/extension glue.
  - CPU: NumPy/Arrow/buffer-compatible views into native CPU execution paths where possible.
- Preserve GAFIME's existing backend semantics and precision modes. The Python surface must not silently change execution precision.
- Reuse the same native kernels / semantic implementation; avoid backend duplication.
- Respect caller stream/device semantics on GPU paths and avoid hidden synchronizations where possible.
- Provide capability/introspection hooks so callers can discover which exported operations/backends/precisions are available.
- Make notebook/cloud use first-class: `pip install` + `import gafime` + existing framework array should be sufficient for the common path.

## Non-goals

- Rewriting GAFIME in CuPy/Torch.
- Creating a second algorithm implementation whose behavior can drift from the native backend.
- Replacing Arrow or existing memory interfaces; this complements them with a **function interface**.
- Exposing every internal helper immediately. Start with a bounded, stable subset.

## Initial candidate

Use one representative native metric/kernel (for example the FP32 MI histogram path) to prove the interface end-to-end:

1. Existing native CUDA implementation remains the source of truth.
2. Export a Python-callable CuPy surface.
3. Accept an already-device-resident CuPy input without host staging.
4. Match native GAFIME results on the same input.
5. Validate stream/device behavior and error propagation.
6. Measure wrapper overhead separately from kernel time.

Then extend the same mechanism to Torch and selected CPU execution functions.

## Acceptance criteria

- [ ] One native CUDA operation is callable directly from CuPy-owned device memory.
- [ ] No duplicate algorithm/kernel implementation is introduced for the CuPy path.
- [ ] Correctness matches the normal native GAFIME path on the same corpus.
- [ ] No mandatory host copy for compatible device buffers.
- [ ] Execution precision is explicit and test-covered.
- [ ] Stream/device ownership and synchronization behavior are documented.
- [ ] Python-facing error handling is deterministic.
- [ ] A minimal Colab example works from a clean runtime.
- [ ] The design is generalized enough to add Torch and CPU function surfaces without inventing a separate architecture for each framework.

## Release target

Candidate for **v1.1** as an integration-focused feature: extend GAFIME's reach into Python GPU/CPU ecosystems without changing its computational core or project identity.
